README.md
App Store Connect Api
====

This is a Python wrapper around the **Apple App Store Api** : https://developer.apple.com/documentation/appstoreconnectapi

So far, it handles token generation / expiration, methods for listing resources and downloading reports. 

Installation
------------

[![Version](http://img.shields.io/pypi/v/appstoreconnect.svg?style=flat)](https://pypi.org/project/appstoreconnect/)

The project is published on PyPI, install with: 

    pip install appstoreconnect

Usage
-----

Please follow instructions on [Apple documentation](https://developer.apple.com/documentation/appstoreconnectapi/creating_api_keys_for_app_store_connect_api) on how to generate an API key.

With your *key ID*, *key file* (you can either pass the path to the file or the content of it as a string) and *issuer ID* create a new API instance:

```python
from appstoreconnect import Api, UserRole
api = Api(key_id, path_to_key_file, issuer_id)

# use a proxy
api = Api(key_id, path_to_key_file, issuer_id, proxy='http://1.2.3.4:3128')

# set a timeout (in seconds) for requests
api = Api(key_id, path_to_key_file, issuer_id, timeout=42)
```

Here are a few examples of API usage. For a complete list of available methods please see [api.py](https://github.com/Ponytech/appstoreconnectapi/blob/master/appstoreconnect/api.py#L148).

```python
# list all apps
apps = api.list_apps()
for app in apps:
    print(app.name, app.sku)

# sort resources
apps = api.list_apps(sort='name')

# filter apps
apps = api.list_apps(filters={'sku': 'DINORUSH', 'name': 'Dino Rush'})
print("%d apps found" % len(apps))

# read app information
app = api.read_app_information('1308363336')
print(app.name, app.sku, app.bundleId)

# get a related resource
for group in app.betaGroups():
    print(group.name)

# list bundle ids
for bundle_id in api.list_bundle_ids():
    print(bundle_id.identifier)

# list certificates
for certificate in api.list_certificates():
    print(certificate.name)

# modify a user
user = api.list_users(filters={'username': 'finance@nemoidstudio.com'})[0]
api.modify_user_account(user, roles=[UserRole.FINANCE, UserRole.ACCESS_TO_REPORTS])
    
# download sales report
api.download_sales_and_trends_reports(
    filters={'vendorNumber': '123456789', 'frequency': 'WEEKLY', 'reportDate': '2019-06-09'}, save_to='report.csv')

# download finance report
api.download_finance_reports(filters={'vendorNumber': '123456789', 'reportDate': '2019-06'}, save_to='finance.csv')
```

Define a timeout (in seconds) after which an exception is raised if no response is received. 

```python
api = Api(key_id, path_to_key_file, issuer_id, timeout=30)
api.list_apps()

APIError: Read timeout after 30 seconds
```


Please note this is a work in progress, API is subject to change between versions.

Anonymous data collection
-------------------------

Starting with version 0.8.0 this library anonymously collects its usage to help better improve its development. 
What we collect is:

- a SHA1 hash of the issuer_id
- the OS and Python version used
- which enpoints had been used

You can review the [source code](https://github.com/Ponytech/appstoreconnectapi/blob/b73d4314e2a9f9098f3287f57fff687563e70b28/appstoreconnect/api.py#L238)

If you feel uncomfortable with it you can completely opt-out by initliazing the API with:

```python
api = Api(key_id, path_to_key_file, issuer_id, submit_stats=False)
```

The is also an [open issue](https://github.com/Ponytech/appstoreconnectapi/issues/18) about this topic where we would love to here your feedback and best practices.


Development
-----------

Project development happens on [Github](https://github.com/Ponytech/appstoreconnectapi) 


TODO
----

* [ ] Support App Store Connect API 1.2
* [ ] Support the include parameter
* [X] handle POST, DELETE and PATCH requests
* [X] sales report
* [X] handle related resources
* [X] allow to sort resources
* [ ] proper API documentation
* [ ] add tests


Credits
-------

This project is developed by [Ponytech](https://ponytech.net)


CHANGELOG.md
## 0.10.1
Bugfixes:
- Relax cryptography dependencies (@conformist-mw)
- Do not assume presence of content-type header (@jaysoffian)
- Fix a possible RecursionError 

## 0.10.0

Features:
- Add a timeout parameter (in seconds) for requests
- Add proxy support

Bugfixes:
- Avoid a RecursionError when accessing an unknown attribute on some resources

## 0.9.1

Bugfixes:
- Relax required dependencies in setup.py 
- Fix APIError exception handling
- Add relationships attribute to the Build resource

## 0.9.0

Features:
- New endpoint: `modify_user_account`
- Support getting more related resources, like `user.visibleApps()`

Bugfixes:
- Pin dependencies versions in setup.py

## 0.8.4

Features:
-  Expose the HTTP status code when App Store Connect API response raises APIError (@GClunies)

## 0.8.3

Bugfixes:
- Fix invite_user method (@AricWu)

## 0.8.2

Features:
 - New `split_response` argument in `download_finance_reports()` function that splits the response into 2 objects. Defualt value is `split_response=False`. This also
 allows the 2 responses to be saved to separate files using a syntax like `save_to=['test1.csv', 'test2.csv']`. (@GClunies)

## 0.8.1

Bugfixes:
 - Add default versions and subtypes in download_sales_and_trends_reports

## 0.8.0

Features:
 - New endpoints:
   - delete_beta_tester
   - read_beta_tester_information
   - modify_beta_group
   - delete_beta_group
   - read_beta_group_information
   - read_beta_app_localization_information
   - create_beta_app_localization
   - modify_registered_device
   - read_beta_app_review_submission_information
 - Collect anonymous usage statistics

Breaking changes API:
 - new parameters for create_beta_tester
 - new parameters for create_beta_group
 - new parameters for submit_app_for_beta_review
 - register_device renamed to register_new_device 

## 0.7.0

Features:
 - New endpoint: register_device (@BalestraPatrick)

## 0.6.0

Features:
 - New endpoints: invite_user and read_user_invitation_information (@BalestraPatrick)

Bugfixes:
 - Fixes create_beta_tester endpoint URL (@BalestraPatrick)

## 0.5.1

Bugfixes:
 - Fixes token re-generation (@gsaraceno)

## 0.5.0

Features:
 -  Handle listing all resources in the provisioning section (devices thanks to @EricG-Personal)

## 0.4.1

Features:
 - Allow to query resources sorted
 - Allow passing key as a string value (@kpotehin)

Bugfixes:
 - Fixed sort param in reports (@kpotehin)

## 0.4.0

Features:
 - Handle fetching related resources (@WangYi)

Bugfixes:
 - When paging resources, fix missing resource in the first page (@WangYi)

## 0.3.0

Features:
  - Complete API rewrite, "list" methods return an iterator over resources, "get" method returns a resource 
  - Handles all GET endpoints (except the new "Provisioning" section)
  - Handle pagination
  - Handle downloading Finance and Sales reports

## 0.2.1

Bugfixes:

  - Cryptography dependency is required

## 0.2.0

Features:

  - Added more functions (@fehmitoumi)

## 0.1.0

Features:

  - Initial Release


AppstoreConnect.py
import requests
import jwt
from datetime import datetime, timedelta
import time

ALGORITHM = 'ES256'
BASE_API = "https://api.appstoreconnect.apple.com"


class Api:

	def __init__(self, key_id, key_file, issuer_id):
		self._token = None
		self.token_gen_date = None
		self.exp = None
		self.key_id = key_id
		self.key_file = key_file
		self.issuer_id = issuer_id
		token = self.token  # generate first token

	def _generate_token(self):
		key = open(self.key_file, 'r').read()
		self.token_gen_date = datetime.now()
		exp = int(time.mktime((self.token_gen_date + timedelta(minutes=20)).timetuple()))
		return jwt.encode({'iss': self.issuer_id, 'exp': exp, 'aud': 'appstoreconnect-v1'}, key,
		                   headers={'kid': self.key_id, 'typ': 'JWT'}, algorithm=ALGORITHM).decode('ascii')

	def _api_call(self, route):
		headers = {"Authorization": "Bearer %s" % self.token}
		url = "%s%s" % (BASE_API, route)
		r = requests.get(url, headers=headers)
		if r.status_code == 200:
			return r.json()

	def apps(self):
		return self._api_call("/v1/apps")

	def users(self):
		return self._api_call("/v1/users")

	def user_invitations(self):
		return self._api_call("/v1/userInvitations")

	def beta_groups(self):
		return self._api_call("/v1/betaGroups")

	def beta_testers(self):
		return self._api_call("/v1/betaTesters")

	def builds(self):
		return self._api_call("/v1/builds")

	@property
	def token(self):
		# generate a new token every 15 minutes
		if not self._token or self.token_gen_date + timedelta(minutes=15) > datetime.now():
			self._token = self._generate_token()

		return self._token

appstoreconnect/resources.py
import inspect
from abc import ABC, abstractmethod
import sys


class Resource(ABC):
	relationships = {}

	def __init__(self, data, api):
		self._data = data
		self._api = api

	def __getattr__(self, item):
		if item == 'id':
			return self._data.get('id')
		if item in self._data.get('attributes', {}):
			return self._data.get('attributes', {})[item]
		if item in self.relationships:
			def getter():
				# Try to fetch relationship
				nonlocal item
				url = self._data.get('relationships', {})[item]['links']['related']
				if self.relationships[item]['multiple']:
					return self._api.get_related_resources(full_url=url)
				else:
					return self._api.get_related_resource(full_url=url)
			return getter

		raise AttributeError('%s has no attributes %s' % (self.type_name, item))

	def __repr__(self):
		return '%s id %s' % (self.type_name, self._data.get('id'))

	def __dir__(self):
		return ['id'] + list(self._data.get('attributes', {}).keys()) + list(self._data.get('relationships', {}).keys())

	@property
	def type_name(self):
		return type(self).__name__

	@property
	@abstractmethod
	def endpoint(self):
		pass


# Beta Testers and Groups

class BetaTester(Resource):
	endpoint = '/v1/betaTesters'
	type = 'betaTesters'
	attributes = ['email', 'firstName', 'inviteType', 'lastName']
	relationships = {
		'apps': {'multiple': True},
		'betaGroups': {'multiple': True},
		'builds': {'multiple': True},
	}
	documentation = 'https://developer.apple.com/documentation/appstoreconnectapi/betatester'

class BetaTesterInvitation(Resource):
	endpoint = '/v1/betaTesterInvitations'
	type = 'betaTesterInvitations'
	attributes = []
	relationships = {
		'app': {'multiple': False},
		'betaTester': {'multiple': False}
	}
	documentation = 'https://developer.apple.com/documentation/appstoreconnectapi/BetaTesterInvitation'

class BetaGroup(Resource):
	endpoint = '/v1/betaGroups'
	type = 'betaGroups'
	attributes = ['isInternalGroup', 'name', 'publicLink', 'publicLinkEnabled', 'publicLinkId', 'publicLinkLimit', 'publicLinkLimitEnabled', 'createdDate']
	relationships = {
		'app': {'multiple': False},
		'betaTesters': {'multiple': True},
		'builds': {'multiple': True},
	}
	documentation = 'https://developer.apple.com/documentation/appstoreconnectapi/betagroup'



# App Metadata Resources

class AppScreenshot(Resource):
	endpoint = '/v1/appScreenshots'
	type = 'appScreenshots'
	attributes = ['assetDeliveryState', 'assetToken', 'assetType', 'fileName', 'fileSize', 'imageAsset', 'sourceFileChecksum', 'uploadOperations', 'uploaded']
	relationships = {
		'appScreenshotSet': {'multiple': False}
	}
	documentation = 'https://developer.apple.com/documentation/appstoreconnectapi/appscreenshot'

class AppScreenshotSet(Resource):
	endpoint = '/v1/appScreenshotSets'
	type = 'appScreenshotSets'
	attributes = ['screenshotDisplayType']
	relationships = {
		'appScreenshots': {'multiple': True},
		'appStoreVersionLocalization': {'multiple': False}
	}
	documentation = 'https://developer.apple.com/documentation/appstoreconnectapi/appscreenshotset'

class AppStoreVersionLocalization(Resource):
	endpoint = '/v1/appStoreVersionLocalizations'
	type = 'appStoreVersionLocalizations'
	attributes = ['description', 'keywords', 'locale', 'marketingUrl', 'promotionalText', 'supportUrl', 'whatsNew']
	relationships = {
		'appPreviewSets': {'multiple': True},
		'appScreenshotSets': {'multiple': True},
		'appStoreVersion': {'multiple': False}
	}
	documentation = 'https://developer.apple.com/documentation/appstoreconnectapi/appstoreversionlocalization'

class AgeRatingDeclarations(Resource):
	endpoint = '/v1/ageRatingDeclarations'
	attributes = ['alcoholTobaccoOrDrugUseOrReferences', 'gamblingAndContests', 'kidsAgeBand', 'medicalOrTreatmentInformation',
	'profanityOrCrudeHumor', 'sexualContentOrNudity', 'unrestrictedWebAccess', 'gamblingSimulated', 'horrorOrFearThemes',
	'matureOrSuggestiveThemes', 'sexualContentGraphicAndNudity', 'violenceCartoonOrFantasy', 'violenceRealistic', 'violenceRealisticProlongedGraphicOrSadistic']
	relationships = {}
	type = 'ageRatingDeclarations'
	documentation = 'https://developer.apple.com/documentation/appstoreconnectapi/ageratingdeclaration'

class AppStoreVersion(Resource):
	endpoint = '/v1/appStoreVersions'
	type = 'appStoreVersions'
	attributes = ['platform', 'appStoreState', 'copyright', 'earliestReleaseDate', 'releaseType', 'usesIdfa', 'versionString', 'downloadable', 'createdDate']
	relationships = {
		'app': {'multiple': False},
		'ageRatingDeclaration': {'multiple': False},
		'appStoreReviewDetail': {'multiple': False},
		'appStoreVersionLocalizations': {'multiple': True},
		'appStoreVersionPhasedRelease': {'multiple': False},
		'appStoreVersionSubmission': {'multiple': False},
		'build': {'multiple': False},
		'idfaDeclaration': {'multiple': False},
		'routingAppCoverage': {'multiple': False},
	}
	documentation = 'https://developer.apple.com/documentation/appstoreconnectapi/appstoreversion'

class AgeRatingDeclaration(Resource):
	endpoint = '/v1/ageRatingDeclarations'
	type = "ageRatingDeclarations"
	documentation = 'https://developer.apple.com/documentation/appstoreconnectapi/ageratingdeclaration'

class Territory(Resource):
	endpoint = '/v1/territories'
	type = 'territories'
	attributes = 'currency'
	relationships = {}
	documentation = 'https://developer.apple.com/documentation/appstoreconnectapi/territory'


class AppStoreReviewDetail(Resource):
	endpoint = '/v1/appStoreReviewDetails'
	type = "appStoreReviewDetails"
	attributes = ['contactEmail', 'contactFirstName', 'contactLastName', 'contactPhone', 'demoAccountName', 'demoAccountPassword', 'demoAccountRequired', 'notes']
	relationships = {
		'appStoreVersion': {'multiple': False},
		'appStoreReviewAttachments': {'multiple': True}
	}
	documentation = 'https://developer.apple.com/documentation/appstoreconnectapi/appstorereviewdetail'

class AppStoreVersionLocalizations(Resource):
	endpoint = '/v1/appStoreVersionLocalizations'
	type = "appStoreVersionLocalizations"
	documentation = 'https://developer.apple.com/documentation/appstoreconnectapi/appstoreversionlocalization'

class AppStoreVersionSubmission(Resource):
	endpoint = '/v1/appStoreVersionSubmissions'
	type = "appStoreVersionSubmissions"
	attributes = []
	relationships = {
		'appStoreVersion': {'multiple': False}
	}
	documentation = 'https://developer.apple.com/documentation/appstoreconnectapi/appstoreversionsubmission'

class IdfaDeclaration(Resource):
	endpoint = '/v1/idfaDeclarations'
	type = "idfaDeclarations"
	documentation = 'https://developer.apple.com/documentation/appstoreconnectapi/idfadeclaration'

class RoutingAppCoverage(Resource):
	endpoint = '/v1/routingAppCoverages'
	type = "routingAppCoverages"
	documentation = 'https://developer.apple.com/documentation/appstoreconnectapi/routingappcoverages'

class AppInfoLocalization(Resource):
	endpoint = '/v1/appInfoLocalizations'
	type = 'appInfoLocalizations'
	attributes = ['locale', 'name', 'privacyPolicyText', 'privacyPolicyUrl', 'subtitle', 'privacyChoicesUrl']
	relationships = {
		'appInfo': {'multiple': False},
	}
	documentation = 'https://developer.apple.com/documentation/appstoreconnectapi/appinfolocalization'

class AppStoreVersionPhasedRelease(Resource):
	endpoint = '/v1/appStoreVersionPhasedReleases'
	type = 'appStoreVersionPhasedReleases'
	attributes = ['currentDayNumber', 'phasedReleaseState', 'startDate', 'totalPauseDuration']
	relationships = {
		'appStoreVersion': {'multiple': False},
	}
	documentation = 'https://developer.apple.com/documentation/appstoreconnectapi/appstoreversionphasedrelease'

# App Resources

class App(Resource):
	endpoint = '/v1/apps'
	type = 'apps'
	attributes = ['bundleId', 'name', 'primaryLocale', 'sku']
	relationships = {
		'betaLicenseAgreement': {'multiple': False},
		'preReleaseVersions': {'multiple': True},
		'betaAppLocalizations': {'multiple': True},
		'betaGroups': {'multiple': True},
		'betaTesters': {'multiple': True},
		'builds': {'multiple': True},
		'betaAppReviewDetail': {'multiple': False},
	}
	documentation = 'https://developer.apple.com/documentation/appstoreconnectapi/app'


class PreReleaseVersion(Resource):
	endpoint = '/v1/preReleaseVersions'
	type = 'preReleaseVersion'
	attributes = ['platform', 'version']
	documentation = 'https://developer.apple.com/documentation/appstoreconnectapi/preReleaseVersion/attributes'


class BetaAppLocalization(Resource):
	endpoint = '/v1/betaAppLocalizations'
	type = 'betaAppLocalizations'
	attributes = ['description', 'feedbackEmail', 'locale', 'marketingUrl', 'privacyPolicyUrl', 'tvOsPrivacyPolicy']
	relationships = {
		'app': {'multiple': False}
	}
	documentation = 'https://developer.apple.com/documentation/appstoreconnectapi/betaAppLocalization/attributes'


class AppEncryptionDeclaration(Resource):
	endpoint = '/v1/appEncryptionDeclarations'
	type = 'appEncryptionDeclarations'
	documentation = 'https://developer.apple.com/documentation/appstoreconnectapi/appEncryptionDeclaration/attributes'


class BetaLicenseAgreement(Resource):
	endpoint = '/v1/betaLicenseAgreements'
	documentation = 'https://developer.apple.com/documentation/appstoreconnectapi/betaLicenseAgreement/attributes'


# Build Resources

class Build(Resource):
	endpoint = '/v1/builds'
	type = 'builds'
	attributes = ['expired', 'iconAssetToken', 'minOsVersion', 'processingState', 'version', 'usesNonExemptEncryption', 'uploadedDate', 'expirationDate']
	relationships = {
		'app': {'multiple': False},
		'appEncryptionDeclaration': {'multiple': False},
		'individualTesters': {'multiple': True},
		'preReleaseVersion': {'multiple': False},
		'betaBuildLocalizations': {'multiple': True},
		'buildBetaDetail': {'multiple': False},
		'betaAppReviewSubmission': {'multiple': False},
		'appStoreVersion': {'multiple': False},
		'icons': {'multiple': True},
		'buildBundles': {'multiple': True},
		'betaGroups': {'multiple': True},
	}
	documentation = 'https://developer.apple.com/documentation/appstoreconnectapi/build/attributes'


class BuildBetaDetail(Resource):
	endpoint = '/v1/buildBetaDetails'
	attributes = ['autoNotifyEnabled', 'externalBuildState', 'internalBuildState']
	relationships = {
		'build': {'multiple': False},
	}
	type = 'buildBetaDetails'
	documentation = 'https://developer.apple.com/documentation/appstoreconnectapi/buildBetaDetail/attributes'


class BetaBuildLocalization(Resource):
	endpoint = '/v1/betaBuildLocalizations'
	type = 'betaBuildLocalizations'
	attributes = ['locale', 'whatsNew']
	relationships = {
		'build': {'multiple': False},
	}
	documentation = 'https://developer.apple.com/documentation/appstoreconnectapi/betaBuildLocalization/attributes'


class BetaAppReviewDetail(Resource):
	endpoint = '/v1/betaAppReviewDetails'
	type = 'betaAppReviewDetails'
	attributes = ['contactEmail', 'contactFirstName', 'contactLastName', 'contactPhone', 'demoAccountName', 'demoAccountPassword', 'demoAccountRequired', 'notes']
	relationships = {
		'app': {'multiple': False},
	}
	documentation = 'https://developer.apple.com/documentation/appstoreconnectapi/betaAppReviewDetail/attributes'


class BetaAppReviewSubmission(Resource):
	endpoint = '/v1/betaAppReviewSubmissions'
	type = 'betaAppReviewSubmissions'
	attributes = ['betaReviewState']
	relationships = {
		'build': {'multiple': False},
	}
	documentation = 'https://developer.apple.com/documentation/appstoreconnectapi/betaAppReviewSubmission/attributes'


# Users and Roles

class User(Resource):
	endpoint = '/v1/users'
	documentation = 'https://developer.apple.com/documentation/appstoreconnectapi/user/attributes'


class UserInvitation(Resource):
	endpoint = '/v1/userInvitations'
	documentation = 'https://developer.apple.com/documentation/appstoreconnectapi/userinvitation/attributes'


# Provisioning
class BundleId(Resource):
	endpoint = '/v1/bundleIds'
	type = 'bundleIds'
	attributes = ['identifier', 'name', 'platform', 'seedId']
	relationships = {
		'profiles': {'multiple': True},
		'bundleIdCapabilities': {'multiple': True},
		'app': {'multiple': False},
	}
	documentation = 'https://developer.apple.com/documentation/appstoreconnectapi/bundleid/attributes'


class Certificate(Resource):
	endpoint = '/v1/certificates'
	documentation = 'https://developer.apple.com/documentation/appstoreconnectapi/certificate/attributes'


class Device(Resource):
	endpoint = '/v1/devices'
	type = 'devices'
	attributes = ['name', 'platform', 'udid', 'status']
	relationships = {
	}
	documentation = 'https://developer.apple.com/documentation/appstoreconnectapi/device/attributes'


class Profile(Resource):
	endpoint = '/v1/profiles'
	documentation = 'https://developer.apple.com/documentation/appstoreconnectapi/profile/attributes'
	type = 'profiles'
	attributes = ['name', 'platform', 'profileContent', 'uuid', 'createdDate', 'profileState', 'profileType', 'expirationDate']
	relationships = {
		'certificates': {'multiple': True},
		'devices': {'multiple': True},
		'bundleId': {'multiple': False},
	}

# Reporting

class FinanceReport(Resource):
	endpoint = '/v1/financeReports'
	filters = 'https://developer.apple.com/documentation/appstoreconnectapi/download_finance_reports'


class SalesReport(Resource):
	endpoint = '/v1/salesReports'
	filters = 'https://developer.apple.com/documentation/appstoreconnectapi/download_sales_and_trends_reports'

class AppCategory(Resource):
	endpoint = '/v1/appCategories'
	type = 'appCategories'
	attributes = ['platforms']
	relationships = {
		'parent': {'multiple': False},
		'subcategories': {'multiple': True}
	}
	documentation = 'https://developer.apple.com/documentation/appstoreconnectapi/appcategory'

class AppInfo(Resource):
	endpoint = '/v1/appInfos'
	type = 'appInfos'
	attributes = ['appStoreAgeRating', 'appStoreState', 'brazilAgeRating', 'kidsAgeBand']
	relationships = {
		'app': {'multiple': False},
		'appInfoLocalizations': {'multiple': True},
		'primaryCategory': {'multiple': False},
		'primarySubcategoryOne': {'multiple': False},
		'primarySubcategoryTwo': {'multiple': False},
		'secondaryCategory': {'multiple': False},
		'secondarySubcategoryOne': {'multiple': False},
		'secondarySubcategoryTwo': {'multiple': False}

	}
	documentation = 'https://developer.apple.com/documentation/appstoreconnectapi/appinfo'


resources = {}
for name, obj in inspect.getmembers(sys.modules[__name__]):
	if inspect.isclass(obj) and issubclass(obj, Resource) and hasattr(obj, 'type') and obj != Resource:
		resources[getattr(obj, 'type')] = obj


example.py
#!/usr/bin/env python

import sys
from appstoreconnect import Api, UserRole


if __name__ == "__main__":
	key_id = sys.argv[1]
	key_file = sys.argv[2]
	issuer_id = sys.argv[3]
	api = Api(key_id, key_file, issuer_id)

	# list all apps
	apps = api.list_apps()
	for app in apps:
		print(app.name, app.sku)

	# filter apps
	apps = api.list_apps(filters={'sku': 'DINORUSH', 'name': 'Dino Rush'})
	print("%d apps found" % len(apps))

	# modify a user
	user = api.list_users(filters={'username': 'finance@nemoidstudio.com'})[0]
	api.modify_user_account(user, roles=[UserRole.FINANCE, UserRole.APP_MANAGER, UserRole.ACCESS_TO_REPORTS])

	# download sales report
	api.download_sales_and_trends_reports(
		filters={'vendorNumber': '123456789', 'frequency': 'WEEKLY', 'reportDate': '2019-06-09'}, save_to='report.csv')

	# download finance report
	api.download_finance_reports(filters={'vendorNumber': '123456789', 'reportDate': '2019-06'}, save_to='finance.csv')
